voltron proxy does not redirect exceptions to clients

use case: try to do a management call to a client that is no more connected to the server.

server crashes.

on server:

```
org.terracotta.exception.EntityUserException: Entity: org.terracotta.management.entity.tms.client.TmsAgentEntity:VoltronTmsConnection:MyCluster exception in user code: org.terracotta.exception.EntityUserException: Entity: org.terracotta.management.entity.tms.client.TmsAgentEntity:VoltronTmsConnection:MyCluster exception in user code: Runtime exception in deserializer
	at com.tc.objectserver.entity.ManagedEntityImpl.invoke(ManagedEntityImpl.java:497)
	at com.tc.objectserver.entity.ManagedEntityImpl.lambda$processInvokeRequest$2(ManagedEntityImpl.java:308)
	at com.tc.objectserver.entity.ManagedEntityImpl$SchedulingRunnable.run(ManagedEntityImpl.java:1025)
	at com.tc.objectserver.entity.RequestProcessor$EntityRequest.invoke(RequestProcessor.java:173)
	at com.tc.objectserver.entity.RequestProcessor$EntityRequest.run(RequestProcessor.java:164)
	at com.tc.objectserver.entity.RequestProcessorHandler.handleEvent(RequestProcessorHandler.java:27)
	at com.tc.objectserver.entity.RequestProcessorHandler.handleEvent(RequestProcessorHandler.java:23)
	at com.tc.async.impl.StageQueueImpl$HandledContext.runWithHandler(StageQueueImpl.java:502)
	at com.tc.async.impl.StageImpl$WorkerThread.run(StageImpl.java:197)
Caused by: java.lang.RuntimeException: org.terracotta.exception.EntityUserException: Entity: org.terracotta.management.entity.tms.client.TmsAgentEntity:VoltronTmsConnection:MyCluster exception in user code: Runtime exception in deserializer
	at com.tc.objectserver.entity.ManagedEntityImpl.performAction(ManagedEntityImpl.java:724)
	at com.tc.objectserver.entity.ManagedEntityImpl.invoke(ManagedEntityImpl.java:472)
	... 8 more
Caused by: org.terracotta.exception.EntityUserException: Entity: org.terracotta.management.entity.tms.client.TmsAgentEntity:VoltronTmsConnection:MyCluster exception in user code: Runtime exception in deserializer
	at com.tc.objectserver.entity.ManagedEntityImpl.runWithHelper(ManagedEntityImpl.java:388)
	at com.tc.objectserver.entity.ManagedEntityImpl.performAction(ManagedEntityImpl.java:717)
	... 9 more
Caused by: org.terracotta.entity.MessageCodecException: Runtime exception in deserializer
	at com.tc.objectserver.entity.ManagedEntityImpl.runWithHelper(ManagedEntityImpl.java:387)
	... 10 more
Caused by: java.lang.IllegalStateException: Client 81279@127.0.0.1:Ehcache:SEFormy-cm-name-2:2b72ff6c02ff43e29246443ba635324a is either not found or not manageable
	at org.terracotta.management.service.monitoring.DefaultManagementService.lambda$sendManagementCallRequest$0(DefaultManagementService.java:105)
	at java.util.Optional.orElseThrow(Optional.java:290)
	at org.terracotta.management.service.monitoring.DefaultManagementService.sendManagementCallRequest(DefaultManagementService.java:105)
	at org.terracotta.management.entity.tms.server.ActiveTmsAgent.call(ActiveTmsAgent.java:88)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.terracotta.voltron.proxy.MethodDescriptor.invoke(MethodDescriptor.java:144)
	at org.terracotta.voltron.proxy.ProxyEntityMessage.invoke(ProxyEntityMessage.java:67)
	at org.terracotta.voltron.proxy.server.ProxyInvoker.invoke(ProxyInvoker.java:50)
	at org.terracotta.voltron.proxy.server.ActiveProxiedServerEntity.invoke(ActiveProxiedServerEntity.java:53)
	at org.terracotta.voltron.proxy.server.ActiveProxiedServerEntity.invoke(ActiveProxiedServerEntity.java:30)
	at com.tc.objectserver.entity.ManagedEntityImpl.lambda$performAction$9(ManagedEntityImpl.java:717)
	at com.tc.objectserver.entity.ManagedEntityImpl.runWithHelper(ManagedEntityImpl.java:382)
	... 10 more
com.tc.exception.TCRuntimeException: Uncaught exception in stage
	at com.tc.async.impl.StageImpl$WorkerThread.run(StageImpl.java:211)
Caused by: java.lang.RuntimeException: org.terracotta.exception.EntityUserException: Entity: org.terracotta.management.entity.tms.client.TmsAgentEntity:VoltronTmsConnection:MyCluster exception in user code: org.terracotta.exception.EntityUserException: Entity: org.terracotta.management.entity.tms.client.TmsAgentEntity:VoltronTmsConnection:MyCluster exception in user code: Runtime exception in deserializer
	at com.tc.objectserver.entity.ManagedEntityImpl.invoke(ManagedEntityImpl.java:499)
	at com.tc.objectserver.entity.ManagedEntityImpl.lambda$processInvokeRequest$2(ManagedEntityImpl.java:308)
	at com.tc.objectserver.entity.ManagedEntityImpl$SchedulingRunnable.run(ManagedEntityImpl.java:1025)
	at com.tc.objectserver.entity.RequestProcessor$EntityRequest.invoke(RequestProcessor.java:173)
	at com.tc.objectserver.entity.RequestProcessor$EntityRequest.run(RequestProcessor.java:164)
	at com.tc.objectserver.entity.RequestProcessorHandler.handleEvent(RequestProcessorHandler.java:27)
	at com.tc.objectserver.entity.RequestProcessorHandler.handleEvent(RequestProcessorHandler.java:23)
	at com.tc.async.impl.StageQueueImpl$HandledContext.runWithHandler(StageQueueImpl.java:502)
	at com.tc.async.impl.StageImpl$WorkerThread.run(StageImpl.java:197)
Caused by: org.terracotta.exception.EntityUserException: Entity: org.terracotta.management.entity.tms.client.TmsAgentEntity:VoltronTmsConnection:MyCluster exception in user code: org.terracotta.exception.EntityUserException: Entity: org.terracotta.management.entity.tms.client.TmsAgentEntity:VoltronTmsConnection:MyCluster exception in user code: Runtime exception in deserializer
	at com.tc.objectserver.entity.ManagedEntityImpl.invoke(ManagedEntityImpl.java:497)
	... 8 more
Caused by: java.lang.RuntimeException: org.terracotta.exception.EntityUserException: En
```

on client:

```
Caused by: org.terracotta.exception.EntityUserException: Entity: org.terracotta.management.entity.tms.client.TmsAgentEntity:VoltronTmsConnection:MyCluster exception in user code: Entity: org.terracotta.management.entity.tms.client.TmsAgentEntity:VoltronTmsConnection:MyCluster exception in user code: org.terracotta.entity.MessageCodecException: Runtime exception in deserializer
	at com.tc.object.InFlightMessage.getWithTimeout(InFlightMessage.java:178) ~[client-runtime-5.2.0-pre13.jar:na]
	at com.tc.object.InFlightMessage.getWithTimeout(InFlightMessage.java:43) ~[client-runtime-5.2.0-pre13.jar:na]
	at com.tc.object.EntityClientEndpointImpl$InvocationBuilderImpl$1.getWithTimeout(EntityClientEndpointImpl.java:173) ~[client-runtime-5.2.0-pre13.jar:na]
	at com.tc.object.EntityClientEndpointImpl$InvocationBuilderImpl$1.getWithTimeout(EntityClientEndpointImpl.java:155) ~[client-runtime-5.2.0-pre13.jar:na]
	at org.terracotta.voltron.proxy.client.VoltronProxyInvocationHandler$ProxiedInvokeFuture.get(VoltronProxyInvocationHandler.java:199) ~[management-client-5.2.0-pre15.jar:na]
	at org.terracotta.management.entity.tms.client.DefaultTmsAgentService.get(DefaultTmsAgentService.java:153) ~[management-client-5.2.0-pre15.jar:na]
	at org.terracotta.management.entity.tms.client.DefaultTmsAgentService.call(DefaultTmsAgentService.java:121) ~[management-client-5.2.0-pre15.jar:na]
	at com.terracottatech.management.voltron.VoltronTmsConnection.doCall(VoltronTmsConnection.java:120) ~[classes/:na]
	... 44 common frames omitted
Caused by: org.terracotta.exception.EntityUserException: Entity: org.terracotta.management.entity.tms.client.TmsAgentEntity:VoltronTmsConnection:MyCluster exception in user code: org.terracotta.entity.MessageCodecException: Runtime exception in deserializer
	at org.terracotta.exception.EntityUserException.writeReplace(EntityUserException.java:49) ~[connection-api-1.2.0-pre12.jar:na]
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[na:1.8.0_121]
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[na:1.8.0_121]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:1.8.0_121]
	at java.lang.reflect.Method.invoke(Method.java:498) ~[na:1.8.0_121]
	at java.io.ObjectStreamClass.invokeWriteReplace(ObjectStreamClass.java:1118) ~[na:1.8.0_121]
	at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1136) ~[na:1.8.0_121]
	at java.io.ObjectOutputStream.writeObject(ObjectOutputStream.java:348) ~[na:1.8.0_121]
	at com.tc.entity.VoltronEntityAppliedResponseImpl.dehydrateValues(VoltronEntityAppliedResponseImpl.java:114) ~[client-runtime-5.2.0-pre13.jar:na]
	at com.tc.net.protocol.tcm.TCMessageImpl.nvToTCByteBufferArray(TCMessageImpl.java:149) ~[client-runtime-5.2.0-pre13.jar:na]
	at com.tc.net.protocol.tcm.TCMessageImpl.dehydrate(TCMessageImpl.java:134) ~[client-runtime-5.2.0-pre13.jar:na]
	at com.tc.net.protocol.tcm.TCMessageImpl.dehydrate(TCMessageImpl.java:128) ~[client-runtime-5.2.0-pre13.jar:na]
	at com.tc.net.protocol.tcm.TCMessageImpl.send(TCMessageImpl.java:383) ~[client-runtime-5.2.0-pre13.jar:na]
	at com.tc.objectserver.handler.ProcessTransactionHandler$1.handleEvent(ProcessTransactionHandler.java:109) ~[na:na]
	at com.tc.objectserver.handler.ProcessTransactionHandler$1.handleEvent(ProcessTransactionHandler.java:104) ~[na:na]
	at com.tc.async.impl.StageQueueImpl$HandledContext.runWithHandler(StageQueueImpl.java:502) ~[client-runtime-5.2.0-pre13.jar:na]
	at com.tc.async.impl.StageImpl$WorkerThread.run(StageImpl.java:197) ~[client-runtime-5.2.0-pre13.jar:na]
Caused by: org.terracotta.exception.ServerSideExceptionWrapper: org.terracotta.entity.MessageCodecException: Runtime exception in deserializer
	at com.tc.objectserver.entity.ManagedEntityImpl.runWithHelper(ManagedEntityImpl.java:387) ~[na:na]
	at com.tc.objectserver.entity.ManagedEntityImpl.performAction(ManagedEntityImpl.java:717) ~[na:na]
	at com.tc.objectserver.entity.ManagedEntityImpl.invoke(ManagedEntityImpl.java:472) ~[na:na]
	at com.tc.objectserver.entity.ManagedEntityImpl.lambda$processInvokeRequest$2(ManagedEntityImpl.java:308) ~[na:na]
	at com.tc.objectserver.entity.ManagedEntityImpl$SchedulingRunnable.run(ManagedEntityImpl.java:1025) ~[na:na]
	at com.tc.objectserver.entity.RequestProcessor$EntityRequest.invoke(RequestProcessor.java:173) ~[na:na]
	at com.tc.objectserver.entity.RequestProcessor$EntityRequest.run(RequestProcessor.java:164) ~[na:na]
	at com.tc.objectserver.entity.RequestProcessorHandler.handleEvent(RequestProcessorHandler.java:27) ~[na:na]
	at com.tc.objectserver.entity.RequestProcessorHandler.handleEvent(RequestProcessorHandler.java:23) ~[na:na]
	... 2 common frames omitted
Caused by: org.terracotta.exception.ServerSideExceptionWrapper: java.lang.IllegalStateException: Client 81279@127.0.0.1:Ehcache:SEFormy-cm-name-2:2b72ff6c02ff43e29246443ba635324a is either not found or not manageable
	at org.terracotta.management.service.monitoring.DefaultManagementService.lambda$sendManagementCallRequest$0(DefaultManagementService.java:105) ~[na:na]
	at java.util.Optional.orElseThrow(Optional.java:290) ~[na:1.8.0_121]
	at org.terracotta.management.service.monitoring.DefaultManagementService.sendManagementCallRequest(DefaultManagementService.java:105) ~[na:na]
	at org.terracotta.management.entity.tms.server.ActiveTmsAgent.call(ActiveTmsAgent.java:88) ~[na:na]
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[na:1.8.0_121]
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[na:1.8.0_121]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:1.8.0_121]
	at java.lang.reflect.Method.invoke(Method.java:498) ~[na:1.8.0_121]
	at org.terracotta.voltron.proxy.MethodDescriptor.invoke(MethodDescriptor.java:144) ~[management-client-5.2.0-pre15.jar:na]
	at org.terracotta.voltron.proxy.ProxyEntityMessage.invoke(ProxyEntityMessage.java:67) ~[management-client-5.2.0-pre15.jar:na]
	at org.terracotta.voltron.proxy.server.ProxyInvoker.invoke(ProxyInvoker.java:50) ~[na:na]
	at org.terracotta.voltron.proxy.server.ActiveProxiedServerEntity.invoke(ActiveProxiedServerEntity.java:53) ~[na:na]
	at org.terracotta.voltron.proxy.server.ActiveProxiedServerEntity.invoke(ActiveProxiedServerEntity.java:30) ~[na:na]
	at com.tc.objectserver.entity.ManagedEntityImpl.lambda$performAction$9(ManagedEntityImpl.java:717) ~[na:na]
	at com.tc.objectserver.entity.ManagedEntityImpl.runWithHelper(ManagedEntityImpl.java:382) ~[na:na]
	... 10 common frames omitted
```

--- a/management/monitoring-service/src/main/java/org/terracotta/management/service/monitoring/DefaultManagementService.java
+++ b/management/monitoring-service/src/main/java/org/terracotta/management/service/monitoring/DefaultManagementService.java
@@ -102,7 +102,7 @@ class DefaultManagementService implements ManagementService, TopologyEventListen
       // handle client call
       ClientIdentifier to = ClientIdentifier.valueOf(context.get(Client.KEY));
       fullContext = context.with(topologyService.getManageableClientContext(to)
-          .orElseThrow(() -> new IllegalStateException("Client " + to + " is either not found or not manageable")));
+          .orElseThrow(() -> new IllegalArgumentException("Client " + to + " is either not found or not manageable")));
     }
 
     if ((context.contains(Server.NAME_KEY) || context.contains(Server.KEY))
@@ -112,7 +112,7 @@ class DefaultManagementService implements ManagementService, TopologyEventListen
       Context entityCtx = (context.contains(ServerEntity.CONSUMER_ID) ?
           topologyService.getManageableEntityContext(serverName, Long.parseLong(context.get(ServerEntity.CONSUMER_ID))) :
           topologyService.getManageableEntityContext(serverName, context.get(ServerEntity.NAME_KEY), context.get(ServerEntity.TYPE_KEY)))
-          .orElseThrow(() -> new IllegalStateException("Server Entity " + context + " is either not found or not manageable"));
+          .orElseThrow(() -> new IllegalArgumentException("Server Entity " + context + " is either not found or not manageable"));
       fullContext = context.with(entityCtx);
     }
 

--- a/management/testing/integration-tests/src/test/java/org/terracotta/management/integration/tests/StripeNamedIT.java
+++ b/management/testing/integration-tests/src/test/java/org/terracotta/management/integration/tests/StripeNamedIT.java
@@ -46,7 +46,7 @@ public class StripeNamedIT extends AbstractSingleTest {
   public void stripe_can_be_named() throws Exception {
     // connects to server
     Properties properties = new Properties();
-    properties.setProperty(ConnectionPropertyNames.CONNECTION_NAME, getClass().getSimpleName());
+    properties.setProperty(ConnectionPropertyNames.CONNECTION_NAME, getClass().getSimpleName() + "-2"); // to have a different name than the client created in the super class
     properties.setProperty(ConnectionPropertyNames.CONNECTION_TIMEOUT, "5000");
     Connection managementConnection = ConnectionFactory.connect(voltron.getConnectionURI(), properties);
 
@@ -61,7 +61,7 @@ public class StripeNamedIT extends AbstractSingleTest {
     String currentTopo = toJson(tmsAgentService.readTopology().toMap()).toString();
     String actual = removeRandomValues(currentTopo);
     String expected = readJson("topology-renamed.json").toString();
-    System.out.println(actual);
+
     assertEquals(expected, actual);
 
     // clear previous notifs

--- a/management/testing/integration-tests/src/test/java/org/terracotta/management/integration/tests/UserErrorsIT.java
+++ b/management/testing/integration-tests/src/test/java/org/terracotta/management/integration/tests/UserErrorsIT.java
@@ -1,0 +1,90 @@
+/*
+ * Copyright Terracotta, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.terracotta.management.integration.tests;
+
+import org.junit.Test;
+import org.terracotta.exception.EntityUserException;
+import org.terracotta.management.entity.tms.client.IllegalManagementCallException;
+import org.terracotta.management.model.call.Parameter;
+import org.terracotta.management.model.cluster.Client;
+import org.terracotta.management.model.cluster.ClientIdentifier;
+import org.terracotta.management.model.cluster.Server;
+import org.terracotta.management.model.cluster.ServerEntity;
+import org.terracotta.management.model.context.Context;
+
+import java.util.concurrent.CancellationException;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeoutException;
+
+import static org.hamcrest.CoreMatchers.equalTo;
+import static org.hamcrest.CoreMatchers.instanceOf;
+import static org.hamcrest.CoreMatchers.is;
+import static org.junit.Assert.assertThat;
+import static org.junit.Assert.fail;
+
+/**
+ * @author Mathieu Carbou
+ */
+public class UserErrorsIT extends AbstractSingleTest {
+
+  @Test
+  public void management_call_to_inexisting_client() throws Exception {
+    Client client = tmsAgentService.readTopology()
+        .clientStream()
+        .filter(e -> e.getName().equals("pet-clinic"))
+        .findFirst()
+        .get();
+
+    // similar to cacheManagerName and cacheName context
+    String badClientId = ClientIdentifier.create(1L, "127.0.0.1", "NAME", "uuid").getClientId();
+    Context context = client.getContext()
+        .with("clientId", badClientId)
+        .with("appName", "pet-clinic")
+        .with("cacheName", "pets");
+
+    try {
+      tmsAgentService.call(context, "CacheCalls", "put", Void.TYPE, new Parameter("pet1"), new Parameter("Cat")).waitForReturn();
+    } catch (TimeoutException | InterruptedException | CancellationException | IllegalManagementCallException e) {
+      fail();
+    } catch (ExecutionException e) {
+      assertThat(e.getCause(), is(instanceOf(EntityUserException.class)));
+      assertThat(e.getCause().getMessage(), equalTo("Entity: org.terracotta.management.entity.tms.server.ActiveTmsAgent: exception in user code: java.lang.IllegalArgumentException: Client 1@127.0.0.1:NAME:uuid is either not found or not manageable"));
+    }
+  }
+
+  @Test
+  public void management_call_to_inexisting_entity() throws Exception {
+    ServerEntity serverEntity = tmsAgentService.readTopology()
+        .activeServerEntityStream()
+        .filter(e -> e.getName().equals("pet-clinic/pets"))
+        .findFirst()
+        .get();
+
+    Context context = serverEntity.getContext()
+        .with("cacheName", "pet-clinic/pets")
+        .with(Server.NAME_KEY, "INEXISTING");
+
+    try {
+      tmsAgentService.call(context, "ServerCacheCalls", "put", Void.TYPE, new Parameter("pet1"), new Parameter("Cat")).waitForReturn();
+    } catch (TimeoutException | InterruptedException | CancellationException | IllegalManagementCallException e) {
+      fail();
+    } catch (ExecutionException e) {
+      assertThat(e.getCause(), is(instanceOf(EntityUserException.class)));
+      assertThat(e.getCause().getMessage(), equalTo("Entity: org.terracotta.management.entity.tms.server.ActiveTmsAgent: exception in user code: java.lang.IllegalArgumentException: Server Entity {stripeId=SINGLE, serverId=testServer0, serverName=INEXISTING, entityId=pet-clinic/pets:org.terracotta.management.entity.sample.client.CacheEntity, entityName=pet-clinic/pets, entityType=org.terracotta.management.entity.sample.client.CacheEntity, consumerId=3, cacheName=pet-clinic/pets} is either not found or not manageable"));
+    }
+  }
+
+}

--- a/management/testing/integration-tests/src/test/resources/topology-renamed.json
+++ b/management/testing/integration-tests/src/test/resources/topology-renamed.json
@@ -551,13 +551,13 @@
   ],
   "clients": [
     {
-      "id": "0@127.0.0.1:StripeNamedIT:<uuid>",
+      "id": "0@127.0.0.1:StripeNamedIT-2:<uuid>",
       "pid": 0,
       "hostAddress": "127.0.0.1",
-      "name": "StripeNamedIT",
+      "name": "StripeNamedIT-2",
       "logicalConnectionUid": "<uuid>",
       "vmId": "0@127.0.0.1",
-      "clientId": "0@127.0.0.1:StripeNamedIT:<uuid>",
+      "clientId": "0@127.0.0.1:StripeNamedIT-2:<uuid>",
       "hostName": "<hostname>",
       "tags": [],
       "connections": [
@@ -571,7 +571,7 @@
           "stripeId": "MY_SUPER_STRIPE",
           "serverId": "testServer0",
           "serverEntityIds": {
-            "StripeNamedIT:org.terracotta.management.entity.tms.client.TmsAgentEntity": 1
+            "StripeNamedIT-2:org.terracotta.management.entity.tms.client.TmsAgentEntity": 1
           }
         }
       ],
@@ -598,7 +598,7 @@
           "stripeId": "MY_SUPER_STRIPE",
           "serverId": "testServer0",
           "serverEntityIds": {
-            "StripeNamedIT-2:org.terracotta.management.entity.tms.client.TmsAgentEntity": 1
+            "StripeNamedIT:org.terracotta.management.entity.tms.client.TmsAgentEntity": 1
           }
         }
       ],

--- a/voltron-proxy/voltron-proxy-client/src/main/java/org/terracotta/voltron/proxy/client/VoltronProxyInvocationHandler.java
+++ b/voltron-proxy/voltron-proxy-client/src/main/java/org/terracotta/voltron/proxy/client/VoltronProxyInvocationHandler.java
@@ -22,6 +22,7 @@ import org.terracotta.entity.EntityResponse;
 import org.terracotta.entity.InvocationBuilder;
 import org.terracotta.entity.InvokeFuture;
 import org.terracotta.exception.EntityException;
+import org.terracotta.exception.EntityUserException;
 import org.terracotta.voltron.proxy.Codec;
 import org.terracotta.voltron.proxy.MessageListener;
 import org.terracotta.voltron.proxy.MessageType;
@@ -150,9 +151,18 @@ class VoltronProxyInvocationHandler implements InvocationHandler {
       return new ProxiedInvokeFuture(builder.invoke());
 
     } else {
-      ProxyEntityResponse proxyEntityResponse = builder.invoke().get();
-      return proxyEntityResponse == null ? null : proxyEntityResponse.getResponse();
+      return getResponse(builder.invoke().get());
     }
+  }
+
+  private static Object getResponse(ProxyEntityResponse proxyEntityResponse) throws EntityUserException {
+    if (proxyEntityResponse == null) {
+      return null;
+    }
+    if (proxyEntityResponse.getMessageType() == MessageType.ERROR) {
+      throw (EntityUserException) proxyEntityResponse.getResponse();
+    }
+    return proxyEntityResponse.getResponse();
   }
 
   private static class ProxiedInvokeFuture implements Future {
@@ -187,7 +197,7 @@ class VoltronProxyInvocationHandler implements InvocationHandler {
     @Override
     public Object get() throws InterruptedException, ExecutionException {
       try {
-        return future.get().getResponse();
+        return getResponse(future.get());
       } catch (EntityException e) {
         throw new ExecutionException(e);
       }
@@ -196,7 +206,7 @@ class VoltronProxyInvocationHandler implements InvocationHandler {
     @Override
     public Object get(final long timeout, final TimeUnit unit) throws InterruptedException, ExecutionException, TimeoutException {
       try {
-        return future.getWithTimeout(timeout, unit).getResponse();
+        return getResponse(future.getWithTimeout(timeout, unit));
       } catch (EntityException e) {
         throw new ExecutionException(e);
       }

--- a/voltron-proxy/voltron-proxy-common/pom.xml
+++ b/voltron-proxy/voltron-proxy-common/pom.xml
@@ -39,6 +39,11 @@
       <artifactId>packaging-support</artifactId>
       <scope>provided</scope>
     </dependency>
+    <dependency>
+      <groupId>org.terracotta</groupId>
+      <artifactId>connection-api</artifactId>
+      <scope>provided</scope>
+    </dependency>
   </dependencies>
 
 </project>

--- a/voltron-proxy/voltron-proxy-common/src/main/java/org/terracotta/voltron/proxy/MessageType.java
+++ b/voltron-proxy/voltron-proxy-common/src/main/java/org/terracotta/voltron/proxy/MessageType.java
@@ -22,5 +22,5 @@ import com.tc.classloader.CommonComponent;
  */
 @CommonComponent
 public enum MessageType {
-  MESSAGE, SYNC, MESSENGER
+  MESSAGE, SYNC, MESSENGER, ERROR
 }

--- a/voltron-proxy/voltron-proxy-common/src/main/java/org/terracotta/voltron/proxy/ProxyEntityResponse.java
+++ b/voltron-proxy/voltron-proxy-common/src/main/java/org/terracotta/voltron/proxy/ProxyEntityResponse.java
@@ -17,6 +17,7 @@ package org.terracotta.voltron.proxy;
 
 import com.tc.classloader.CommonComponent;
 import org.terracotta.entity.EntityResponse;
+import org.terracotta.exception.EntityUserException;
 
 /**
  * @author cdennis
@@ -38,6 +39,10 @@ public final class ProxyEntityResponse implements EntityResponse {
 
   public static ProxyEntityResponse messengerResponse(Class<?> responseType, Object response) {
     return response(MessageType.MESSENGER, responseType, response);
+  }
+
+  public static ProxyEntityResponse error(EntityUserException error) {
+    return response(MessageType.ERROR, EntityUserException.class, error);
   }
 
   private final MessageType messageType;

--- a/voltron-proxy/voltron-proxy-common/src/main/java/org/terracotta/voltron/proxy/ProxyMessageCodec.java
+++ b/voltron-proxy/voltron-proxy-common/src/main/java/org/terracotta/voltron/proxy/ProxyMessageCodec.java
@@ -17,6 +17,7 @@ package org.terracotta.voltron.proxy;
 
 import org.terracotta.entity.MessageCodec;
 import org.terracotta.entity.MessageCodecException;
+import org.terracotta.exception.EntityUserException;
 
 import java.io.ByteArrayOutputStream;
 import java.io.DataOutputStream;
@@ -80,24 +81,12 @@ public class ProxyMessageCodec implements MessageCodec<ProxyEntityMessage, Proxy
     if (r == null) {
       return new byte[0];
     }
-
     MessageType messageType = r.getMessageType();
-
-    Map<Class<?>, Byte> mapping = responseMappings.get(messageType);
-    if (mapping == null) {
-      throw new AssertionError("WAT, no mapping for " + messageType);
-    }
-
-    Byte messageTypeIdentifier = mapping.get(r.getResponseType());
-    if (messageTypeIdentifier == null) {
-      throw new AssertionError("WAT, no mapping for " + r.getResponseType().getName());
-    }
-
     ByteArrayOutputStream byteOut = new ByteArrayOutputStream();
     DataOutputStream output = new DataOutputStream(byteOut);
     try {
       output.writeByte(messageType.ordinal());
-      output.writeByte(messageTypeIdentifier);
+      output.writeByte(messageType == MessageType.ERROR ? 0 : getMessageTypeIdentifier(r));
       output.write(codec.encode(r.getResponseType(), r.getResponse()));
       output.close();
     } catch (IOException e) {
@@ -111,19 +100,8 @@ public class ProxyMessageCodec implements MessageCodec<ProxyEntityMessage, Proxy
     if (buffer.length == 0) {
       return null;
     }
-
     MessageType messageType = MessageType.values()[buffer[0]];
-
-    Map<Byte, Class<?>> mapping = reverseResponseMappings.get(messageType);
-    if (mapping == null) {
-      throw new AssertionError("WAT, no mapping for " + messageType);
-    }
-
-    Class<?> responseType = mapping.get(buffer[1]);
-    if (responseType == null) {
-      throw new AssertionError("WAT, no mapping for method " + buffer[1] + " for messageType " + messageType);
-    }
-
+    Class<?> responseType = messageType == MessageType.ERROR ? EntityUserException.class : getResponseType(messageType, buffer[1]);
     Object o = codec.decode(responseType, buffer, 2, buffer.length - 2);
     return ProxyEntityResponse.response(messageType, responseType, o);
   }
@@ -133,16 +111,7 @@ public class ProxyMessageCodec implements MessageCodec<ProxyEntityMessage, Proxy
     try {
       MessageType messageType = message.getType();
       MethodDescriptor method = message.getMethod();
-
-      Map<MethodDescriptor, Byte> mapping = reverseMethodMappings.get(messageType);
-      if (mapping == null) {
-        throw new AssertionError("WAT, no mapping for " + messageType);
-      }
-
-      Byte methodIdentifier = mapping.get(method);
-      if (methodIdentifier == null) {
-        throw new AssertionError("WAT, no mapping for " + method.toGenericString());
-      }
+      Byte methodIdentifier = getMethodIdentifier(message);
 
       Object[] args = message.getArguments();
       final Annotation[][] parameterAnnotations = method.getParameterAnnotations();
@@ -172,18 +141,64 @@ public class ProxyMessageCodec implements MessageCodec<ProxyEntityMessage, Proxy
   @Override
   public ProxyEntityMessage decodeMessage(final byte[] buffer) {
     MessageType messageType = MessageType.values()[buffer[0]];
+    MethodDescriptor method = getMethod(messageType, buffer[1]);
+    return new ProxyEntityMessage(method, codec.decode(method.getParameterTypes(), buffer, 2, buffer.length - 2), messageType);
+  }
 
+  private MethodDescriptor getMethod(MessageType messageType, Byte b) {
     Map<Byte, MethodDescriptor> mapping = methodMappings.get(messageType);
     if (mapping == null) {
-      throw new AssertionError("WAT, no mapping for " + messageType);
+      throw new AssertionError("No mapping for " + messageType);
     }
 
-    MethodDescriptor method = mapping.get(buffer[1]);
+    MethodDescriptor method = mapping.get(b);
     if (method == null) {
-      throw new AssertionError("WAT, no mapping for method " + buffer[1] + " for messageType " + messageType);
+      throw new AssertionError("No mapping for method " + b + " for messageType " + messageType);
     }
 
-    return new ProxyEntityMessage(method, codec.decode(method.getParameterTypes(), buffer, 2, buffer.length - 2), messageType);
+    return method;
+  }
+
+  private Class<?> getResponseType(MessageType messageType, Byte b) {
+    Map<Byte, Class<?>> mapping = reverseResponseMappings.get(messageType);
+    if (mapping == null) {
+      throw new AssertionError("No mapping for " + messageType);
+    }
+
+    Class<?> responseType = mapping.get(b);
+    if (responseType == null) {
+      throw new AssertionError("No mapping for method " + b + " for messageType " + messageType);
+    }
+
+    return responseType;
+  }
+
+  private Byte getMethodIdentifier(ProxyEntityMessage message) {
+    Map<MethodDescriptor, Byte> mapping = reverseMethodMappings.get(message.getType());
+    if (mapping == null) {
+      throw new AssertionError("No mapping for " + message.getType());
+    }
+
+    Byte methodIdentifier = mapping.get(message.getMethod());
+    if (methodIdentifier == null) {
+      throw new AssertionError("No mapping for " + message.getMethod().toGenericString());
+    }
+
+    return methodIdentifier;
+  }
+
+  private Byte getMessageTypeIdentifier(ProxyEntityResponse response) {
+    Map<Class<?>, Byte> mapping = responseMappings.get(response.getMessageType());
+    if (mapping == null) {
+      throw new AssertionError("No mapping for " + response.getMessageType());
+    }
+
+    Byte messageTypeIdentifier = mapping.get(response.getResponseType());
+    if (messageTypeIdentifier == null) {
+      throw new AssertionError("No mapping for " + response.getResponseType().getName());
+    }
+
+    return messageTypeIdentifier;
   }
 
 }

--- a/voltron-proxy/voltron-proxy-server/pom.xml
+++ b/voltron-proxy/voltron-proxy-server/pom.xml
@@ -49,6 +49,11 @@
       <artifactId>standard-cluster-services</artifactId>
       <scope>provided</scope>
     </dependency>
+    <dependency>
+      <groupId>org.terracotta</groupId>
+      <artifactId>connection-api</artifactId>
+      <scope>provided</scope>
+    </dependency>
 
     <dependency>
       <groupId>org.terracotta.voltron.proxy</groupId>

--- a/voltron-proxy/voltron-proxy-server/src/main/java/org/terracotta/voltron/proxy/server/ProxyInvoker.java
+++ b/voltron-proxy/voltron-proxy-server/src/main/java/org/terracotta/voltron/proxy/server/ProxyInvoker.java
@@ -18,6 +18,7 @@ package org.terracotta.voltron.proxy.server;
 import org.terracotta.entity.ClientCommunicator;
 import org.terracotta.entity.ClientDescriptor;
 import org.terracotta.entity.MessageCodecException;
+import org.terracotta.exception.EntityUserException;
 import org.terracotta.voltron.proxy.ProxyEntityMessage;
 import org.terracotta.voltron.proxy.ProxyEntityResponse;
 
@@ -45,12 +46,8 @@ class ProxyInvoker<T> implements MessageFiring {
 
   ProxyEntityResponse invoke(final ProxyEntityMessage message, final ClientDescriptor clientDescriptor) {
     try {
-      try {
-        invocationContext.set(new InvocationContext(clientDescriptor));
-        return ProxyEntityResponse.response(message.getType(), message.messageType(), message.invoke(target, clientDescriptor));
-      } finally {
-        invocationContext.remove();
-      }
+      invocationContext.set(new InvocationContext(clientDescriptor));
+      return ProxyEntityResponse.response(message.getType(), message.messageType(), message.invoke(target, clientDescriptor));
     } catch (IllegalAccessException e) {
       throw new IllegalArgumentException(e);
     } catch (InvocationTargetException e) {
@@ -58,10 +55,10 @@ class ProxyInvoker<T> implements MessageFiring {
       if (target instanceof Error) {
         throw (Error) target;
       }
-      if (target instanceof RuntimeException) {
-        throw (RuntimeException) target;
-      }
-      throw new RuntimeException(target.getMessage(), target);
+      EntityUserException entityUserException = new EntityUserException(this.target.getClass().getName(), "", target);
+      return ProxyEntityResponse.error(entityUserException);
+    } finally {
+      invocationContext.remove();
     }
   }
 

